packages/cli/src/commands/hydrogen/deploy.ts

```diff
@@ -0,0 +1,218 @@
+    metadataVersion,
+  } = options;
+  const token = await getOxygenDeploymentToken({
+    root: path,
+    flagShop: shop,
+  });
+  if (!token) {
+    throw new AbortError('Could not obtain Oxygen deployment token');
+  }
+
+  const config: DeploymentConfig = {
+    assetsDir: 'dist/client',
+    deploymentUrl: 'https://oxygen.shopifyapps.com',
+    deploymentToken: parseToken(token as string),
+    healthCheckMaxDuration: 10,
+    metadata: {
+      url: metadataUrl,
+      user: metadataUser,
+      version: metadataVersion,
+    },
+    publicDeployment: publicDeployment,
+    skipHealthCheck: false,
+    rootPath: path,
+    skipBuild: false,
+    workerOnly: false,
+    workerDir: 'dist/worker',
+  };
+
+  let resolveUpload: () => void;
+  const uploadPromise = new Promise<void>((resolve) => {
+    resolveUpload = resolve;
+  });
+
+  let resolveHealthCheck: () => void;
+  const healthCheckPromise = new Promise<void>((resolve) => {
+    resolveHealthCheck = resolve;
+  });
+
+  let rejectDeploy: (reason?: any) => void;
+  const deployPromise = new Promise<void>((_, reject) => {
+    rejectDeploy = reject;
+  });
+
+  const hooks: DeploymentHooks = {
+    buildFunction: async (assetPath: string | undefined): Promise<void> => {
+      outputInfo(
+        outputContent`${colors.whiteBright('Building project...')}`.value,
+      );
+      await runBuild({
+        path,
+        assetPath,
+        sourcemap: false,
+        useCodegen: false,
+      });
+    },
+    onHealthCheckComplete: () => resolveHealthCheck(),
+    onUploadFilesStart: () => uploadStart(),
+    onUploadFilesComplete: () => resolveUpload(),
+    onHealthCheckError: (error: Error) => {
+      rejectDeploy(
+        new AbortError(
+          error.message,
+          'Please verify the deployment status in the Shopify Admin and retry deploying if necessary.',
+        ),
+      );
+    },
+    onUploadFilesError: (error: Error) => {
+      rejectDeploy(
+        new AbortError(
+          error.message,
+          'Check your connection and try again. If the problem persists, try again later or contact support.',
+        ),
+      );
+    },
+  };
+
+  const uploadStart = async () => {
+    outputInfo(
+      outputContent`${colors.whiteBright('Deploying to Oxygen..\n')}`.value,
+    );
+    await renderTasks([
+      {
+        title: 'Uploading files',
+        task: async () => await uploadPromise,
+      },
+      {
+        title: 'Performing health check',
+        task: async () => await healthCheckPromise,
+      },
+    ]);
+  };
+
+  createDeploy({config, hooks, logger: deploymentLogger}).then(
+    (url: string | undefined) => {
+      const deploymentType = config.publicDeployment ? 'public' : 'private';
+      renderSuccess({
+        body: ['Successfully deployed to Oxygen'],
+        nextSteps: [
+          [
+            `Open ${url!} in your browser to view your ${deploymentType} deployment`,
+          ],
+        ],
+      });
+    },
+  );
```

Comment:
I'm not sure how is `createDeploy` implemented, but if it throws an unexpected error (other than health-check / upload-files), this will turn into an unhandled promise rejection. Should we add `.catch(error => rejectDeploy(error))` here just in case?